ratis-examples/src/test/java/org/apache/ratis/examples/filestore/FileStoreBaseTest.java

```diff
@@ -66,14 +62,53 @@
 
   static final int NUM_PEERS = 3;
 
+  FileStoreClient newFileStoreClient(CLUSTER cluster) throws IOException {
+    return new FileStoreClient(cluster.getGroup(), getProperties());
+  }
+
+  @Test
+  public void testWatch() throws Exception {
+    runWithNewCluster(NUM_PEERS, this::runTestWatch);
+  }
+
+  void runTestWatch(CLUSTER cluster) throws Exception {
+    RaftTestUtil.waitForLeader(cluster);
+
+    LOG.info("XXX testWatch");
+    final String path = "testWatch";
+    final AtomicBoolean isStarted = new AtomicBoolean();
+
+    final CompletableFuture<ReadReplyProto> watchFuture;
+    try(FileStoreClient client = new FileStoreClient(cluster.getGroup(), getProperties())) {
+      LOG.info("XXX watchAsync {}", path);
+      watchFuture = client.watchAsync(path).whenComplete((reply, e) -> {
+        Assert.assertNotNull(reply);
+        Assert.assertNull(e);
+        Assert.assertTrue(isStarted.get());
+      });
+      Assert.assertFalse(watchFuture.isDone());
+      Assert.assertFalse(isStarted.get());
+
+      TimeDuration.valueOf(ThreadLocalRandom.current().nextLong(500) + 100, TimeUnit.MILLISECONDS)
+          .sleep(s -> LOG.info("XXX {}", s));
+      Assert.assertFalse(watchFuture.isDone());
+      Assert.assertFalse(isStarted.get());
+      isStarted.set(true);
+      testSingleFile(path, SizeInBytes.ONE_KB, () -> newFileStoreClient(cluster));
```

Comment:
May we test if a read-only request can be processed here (the watch request is applied unordered and thus should not block the following read)?

Reply:
Yes, we should.